// @dynamic-labs/sdk-react-core@4.67.2/src/lib/utils/hooks/useStepUpAuthentication/useStepUpAuthentication.js
'use client'
import { __awaiter } from '../../../../../_virtual/_tslib.js';
import { useState, useRef, useCallback } from 'react';
import { verifyOTP, authenticatePasskeyMFA, authenticateTotpMfaDevice, authenticateMfaRecoveryCode, onEvent, sendEmailOTP, supportedCountries, sendSmsOTP } from '@dynamic-labs-sdk/client';
import { MFADeviceType, JwtVerifiedCredentialFormatEnum } from '@dynamic-labs/sdk-api-core';
import { DeferredPromise } from '@dynamic-labs/utils';
import '@dynamic-labs-sdk/client/core';
import { useDynamicClient } from '../../../client/client.js';
import '../../../config/ApiEndpoint.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import 'react/jsx-runtime';
import { useViewContext } from '../../../context/ViewContext/ViewContext.js';
import { logger } from '../../../shared/logger.js';
import '@dynamic-labs/wallet-book';
import '../../constants/colors.js';
import '../../constants/values.js';
import '../../../shared/consts/index.js';
import '@dynamic-labs/multi-wallet';
import 'react-international-phone';
import '../../../store/state/nonce/nonce.js';
import '@dynamic-labs/locale';
import { useEnvironmentId } from '../../../store/state/dynamicContextProps/dynamicContextProps.js';
import '../../../store/state/primaryWalletId/primaryWalletId.js';
import '../../../store/state/connectedWalletsInfo/connectedWalletsInfo.js';
import '../../functions/getWaasAddressTypeLabel/getWaasAddressTypeLabel.js';
import { dynamicEvents } from '../../../events/dynamicEvents.js';
import { useUser } from '../../../client/extension/user/useUser/useUser.js';
import { hasElevatedAccessToken } from '../../../client/extension/functions/hasElevatedAccessToken/hasElevatedAccessToken.js';
import '../../../context/DynamicContext/DynamicContext.js';
import '../../../store/state/loadingAndLifecycle/loadingAndLifecycle.js';
import { useConnectAndSign } from '../authenticationHooks/useConnectAndSign/useConnectAndSign.js';
import '../../../context/ErrorContext/ErrorContext.js';
import '../../../context/AccessDeniedContext/AccessDeniedContext.js';
import '../../../context/AccountExistsContext/AccountExistsContext.js';
import { useInternalUserWallets } from '../../../context/UserWalletsContext/UserWalletsContext.js';
import '../../../store/state/authMode/authMode.js';
import '../../../context/CaptchaContext/CaptchaContext.js';
import '../../../context/VerificationContext/VerificationContext.js';
import 'react-dom';
import '../../functions/compareChains/compareChains.js';
import '../../../views/Passkey/utils/findPrimaryEmbeddedChain/findPrimaryEmbeddedChain.js';
import '../../../context/ThemeContext/ThemeContext.js';
import '../useUserUpdateRequest/useUpdateUser/userFieldsSchema.js';
import 'bs58';
import '@dynamic-labs/types';
import '../../../context/SocialRedirectContext/SocialRedirectContext.js';
import '../../../context/LoadingContext/LoadingContext.js';
import '../../../context/WalletContext/WalletContext.js';
import '../useEmbeddedWallet/useSecureEnclaveEmbeddedWallet/constants.js';
import 'yup';
import '../../../context/MockContext/MockContext.js';
import '../../../views/CollectUserDataView/useFields.js';
import '../../../context/FieldsStateContext/FieldsStateContext.js';
import '../../../context/UserFieldEditorContext/UserFieldEditorContext.js';
import '@dynamic-labs/rpc-providers';
import '../../../store/state/walletOptions/walletOptions.js';
import 'react-i18next';
import '../../../components/Accordion/components/AccordionItem/AccordionItem.js';
import '../../../components/Alert/Alert.js';
import '../../../components/ShadowDOM/ShadowDOM.js';
import '../../../components/IconButton/IconButton.js';
import '../../../components/InlineWidget/InlineWidget.js';
import '../../../components/Input/Input.js';
import '../../../components/IsBrowser/IsBrowser.js';
import '../../../components/MenuList/Dropdown/Dropdown.js';
import '../../../components/OverlayCard/OverlayCard.js';
import '../../../components/Transition/ZoomTransition/ZoomTransition.js';
import '../../../components/Transition/SlideInUpTransition/SlideInUpTransition.js';
import '../../../components/Transition/OpacityTransition/OpacityTransition.js';
import '../../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.js';
import '../../../components/Popper/Popper/Popper.js';
import '../../../components/Popper/PopperContext/PopperContext.js';
import 'react-focus-lock';
import 'qrcode';
import 'formik';
import '../useSubdomainCheck/useSubdomainCheck.js';
import '../../../context/WalletGroupContext/WalletGroupContext.js';
import '../../../widgets/DynamicWidget/context/DynamicWidgetContext.js';
import '../useWalletBackup/useWalletBackup.js';
import '../useWalletBackup/types.js';
import '../useWalletBackup/cloudProviders.js';
import '../../../context/IpConfigurationContext/IpConfigurationContext.js';
import '../../../context/ConnectWithOtpContext/ConnectWithOtpContext.js';
import '../../../widgets/DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.js';
import '@hcaptcha/react-hcaptcha';
import '../../../widgets/DynamicWidget/helpers/convertExchangeKeyAndProviderEnum.js';
import '../../../views/ExchangeWhitelistWarning/ExchangeWhitelistWarning.js';
import '../../../context/ErrorContext/hooks/useErrorText/useErrorText.js';
import '../../../context/FooterAnimationContext/index.js';
import '../../../views/MfaChooseDeviceView/useGetMfaOptions/useGetMfaOptions.js';
import '../../../context/PasskeyContext/PasskeyContext.js';
import '../../../context/OnrampContext/OnrampContext.js';
import '../../../store/state/sendBalances.js';
import '../../../store/state/connectorsInitializing/connectorsInitializing.js';
import '../../../components/OverlayCardBase/OverlayCardTarget/OverlayCardTarget.js';
import '../../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.js';
import '../../../views/TransactionConfirmationView/TransactionConfirmationView.js';
import '../../../widgets/DynamicWidget/components/PasskeyCard/PasskeyCard.js';
import '../../../widgets/DynamicWidget/views/CryptoComOnramp/CryptoComOnramp.js';
import '../../../../index.js';
import '../../../widgets/DynamicWidget/views/ReceiveWalletFunds/ReceiveWalletFunds.js';
import '../../../store/state/tokenBalances.js';
import '../../../store/state/multichainBalances.js';
import { useGetUserMfaMethods } from '../useGetUserMfaMethods/useGetUserMfaMethods.js';
import { useAuthenticatePasskeyMFA } from '../useAuthenticatePasskeyMFA/useAuthenticatePasskeyMFA.js';
import '../../../shared/utils/functions/getInitialUrl/getInitialUrl.js';
import { useInternalDynamicContext } from '../../../context/DynamicContext/useDynamicContext/useInternalDynamicContext/useInternalDynamicContext.js';

const INITIAL_STATE = {
    error: null,
    isLoading: false,
    otpVerification: null,
};
const isCountryCode = (value) => value !== undefined && value in supportedCountries;
const isOtpCapableCredential = (cred) => cred.format === JwtVerifiedCredentialFormatEnum.Email ||
    cred.format === JwtVerifiedCredentialFormatEnum.PhoneNumber;
const resolveTargetCredential = (credentials, credentialId) => {
    var _a, _b;
    if (credentialId) {
        return ((_a = credentials.find((cred) => cred.id === credentialId)) !== null && _a !== void 0 ? _a : `Credential with id "${credentialId}" not found`);
    }
    return ((_b = credentials.find((cred) => cred.signInEnabled && isOtpCapableCredential(cred))) !== null && _b !== void 0 ? _b : 'No sign-in enabled email or SMS credential found for step-up authentication');
};
const sendEmailOtp = (credential) => __awaiter(void 0, void 0, void 0, function* () {
    if (!credential.email) {
        throw new Error('Email credential is missing email address');
    }
    return sendEmailOTP({ email: credential.email });
});
const sendSmsOtp = (credential) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const normalizedPhoneCountryCode = (_a = credential.phoneCountryCode) === null || _a === void 0 ? void 0 : _a.replace(/^\+/, '');
    const matchedCountryCode = Object.keys(supportedCountries).find((key) => supportedCountries[key].code === normalizedPhoneCountryCode);
    const isoCountryCode = isCountryCode(credential.isoCountryCode)
        ? credential.isoCountryCode
        : matchedCountryCode;
    if (!credential.phoneNumber || !isoCountryCode) {
        throw new Error('SMS credential is missing phone number or country code');
    }
    return sendSmsOTP({ isoCountryCode, phoneNumber: credential.phoneNumber });
});
const dispatchOtp = (credential) => __awaiter(void 0, void 0, void 0, function* () {
    return credential.format === JwtVerifiedCredentialFormatEnum.Email
        ? sendEmailOtp(credential)
        : sendSmsOtp(credential);
});
/**
 * Perform step-up authentication for an already-authenticated user.
 *
 * Supports multiple verification methods:
 * - **OTP** (email/SMS): `sendOtp()` then `verifyOtp()`
 * - **Wallet**: `verifyWallet()` for wallet signature verification
 * - **Passkey MFA**: `verifyPasskeyMfa()` for passkey-based MFA
 * - **TOTP MFA**: `verifyTotpMfa()` for authenticator app codes
 *
 * For OTP and wallet methods, automatically selects the first sign-in enabled
 * credential, or targets a specific one when `credentialId` is provided.
 *
 * @param params.credentialId - Optional id of a specific verified credential or
 *   wallet to use. For OTP, must be an email or SMS credential. Defaults to the
 *   first sign-in enabled email or SMS credential on the authenticated user.
 *
 * @example
 * ```tsx
 * // OTP flow
 * const { sendOtp, verifyOtp } = useStepUpAuthentication();
 * await sendOtp();
 * const result = await verifyOtp({ verificationToken: '123456', requestedScopes: ['wallet:export'] });
 *
 * // Passkey MFA flow
 * const { verifyPasskeyMfa } = useStepUpAuthentication();
 * const result = await verifyPasskeyMfa({ requestedScopes: ['wallet:export'] });
 *
 * // TOTP MFA flow
 * const { verifyTotpMfa } = useStepUpAuthentication();
 * const result = await verifyTotpMfa({ code: '123456', requestedScopes: ['wallet:export'] });
 * ```
 */
const useStepUpAuthentication = ({ credentialId, } = {}) => {
    const user = useUser();
    const environmentId = useEnvironmentId();
    const [state, setState] = useState(INITIAL_STATE);
    const connectAndSign = useConnectAndSign();
    const { userWallets } = useInternalUserWallets();
    const { setShowAuthFlow } = useInternalDynamicContext();
    const { pushView } = useViewContext();
    const getUserMfaMethods = useGetUserMfaMethods();
    const authenticatePasskeyMFA$1 = useAuthenticatePasskeyMFA();
    const dynamicClient = useDynamicClient();
    // Ref keeps the latest otpVerification available inside the verifyOtp
    // callback without creating a stale closure over state.
    const otpVerificationRef = useRef(null);
    const isStepUpRequired = useCallback(({ scope }) => !hasElevatedAccessToken(scope), []);
    const resetState = useCallback(() => {
        otpVerificationRef.current = null;
        setState(INITIAL_STATE);
    }, []);
    const setError = useCallback((errorMessage) => {
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
        logger.error(errorMessage, { environmentId, userId: user === null || user === void 0 ? void 0 : user.id });
    }, [environmentId, user === null || user === void 0 ? void 0 : user.id]);
    const sendOtp = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        const credentials = user === null || user === void 0 ? void 0 : user.verifiedCredentials;
        if (!(credentials === null || credentials === void 0 ? void 0 : credentials.length)) {
            setError('No verified credentials found for the authenticated user');
            return null;
        }
        const result = resolveTargetCredential(credentials, credentialId);
        if (typeof result === 'string') {
            setError(result);
            return null;
        }
        if (!isOtpCapableCredential(result)) {
            setError(`Credential format "${result.format}" does not support OTP verification. Only email and SMS credentials are supported`);
            return null;
        }
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        try {
            const otpVerification = yield dispatchOtp(result);
            otpVerificationRef.current = otpVerification;
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: false, otpVerification })));
            return otpVerification;
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Failed to send OTP';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to send OTP for step-up authentication', {
                environmentId,
                error,
                userId: user === null || user === void 0 ? void 0 : user.id,
            });
            return null;
        }
    }), [
        credentialId,
        environmentId,
        setError,
        user === null || user === void 0 ? void 0 : user.id,
        user === null || user === void 0 ? void 0 : user.verifiedCredentials,
    ]);
    const verifyOtp = useCallback((_a) => __awaiter(void 0, [_a], void 0, function* ({ requestedScopes, verificationToken, }) {
        const otpVerification = otpVerificationRef.current;
        if (!otpVerification) {
            setError('OTP verification has not been initiated. Call sendOtp() first');
            return null;
        }
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        try {
            const response = yield verifyOTP({
                otpVerification,
                requestedScopes,
                verificationToken,
            });
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: false })));
            return response;
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Failed to verify OTP';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to verify OTP for step-up authentication', {
                environmentId,
                error,
                userId: user === null || user === void 0 ? void 0 : user.id,
            });
            return null;
        }
    }), [environmentId, setError, user === null || user === void 0 ? void 0 : user.id]);
    const verifyWallet = useCallback((_b) => __awaiter(void 0, [_b], void 0, function* ({ requestedScopes }) {
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        const wallet = userWallets.find((w) => w.id === credentialId);
        if (!wallet) {
            setError(`Wallet with id "${credentialId}" not found`);
            return;
        }
        try {
            yield connectAndSign({
                requestedScopes,
                walletConnector: wallet.connector,
            });
            setState((prev) => (Object.assign(Object.assign({}, prev), { isLoading: false })));
        }
        catch (error) {
            const errorMessage = error instanceof Error
                ? error.message
                : 'Failed to verify wallet for step-up authentication';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to verify wallet for step-up authentication', {
                environmentId,
                error,
                userId: user === null || user === void 0 ? void 0 : user.id,
            });
        }
    }), [connectAndSign, credentialId, environmentId, setError, user, userWallets]);
    const verifyPasskeyMfa = useCallback((_c) => __awaiter(void 0, [_c], void 0, function* ({ requestedScopes, }) {
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        try {
            const response = yield authenticatePasskeyMFA({
                requestedScopes,
            });
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: false })));
            return response;
        }
        catch (error) {
            const errorMessage = error instanceof Error
                ? error.message
                : 'Failed to authenticate passkey for step-up authentication';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to authenticate passkey for step-up authentication', { environmentId, error, userId: user === null || user === void 0 ? void 0 : user.id });
            return null;
        }
    }), [environmentId, user]);
    const verifyTotpMfa = useCallback((_d) => __awaiter(void 0, [_d], void 0, function* ({ code, deviceId, requestedScopes, }) {
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        try {
            const response = yield authenticateTotpMfaDevice({
                code,
                deviceId,
                requestedScopes,
            });
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: false })));
            return response;
        }
        catch (error) {
            const errorMessage = error instanceof Error
                ? error.message
                : 'Failed to authenticate TOTP device for step-up authentication';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to authenticate TOTP device for step-up authentication', { environmentId, error, userId: user === null || user === void 0 ? void 0 : user.id });
            return null;
        }
    }), [environmentId, user]);
    const verifyRecoveryCode = useCallback((_e) => __awaiter(void 0, [_e], void 0, function* ({ code, requestedScopes, }) {
        setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: true })));
        try {
            const response = yield authenticateMfaRecoveryCode({
                code,
                requestedScopes,
            });
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: null, isLoading: false })));
            return response;
        }
        catch (error) {
            const errorMessage = error instanceof Error
                ? error.message
                : 'Failed to authenticate recovery code for step-up authentication';
            setState((prev) => (Object.assign(Object.assign({}, prev), { error: errorMessage, isLoading: false })));
            logger.error('Failed to authenticate recovery code for step-up authentication', { environmentId, error, userId: user === null || user === void 0 ? void 0 : user.id });
            return null;
        }
    }), [environmentId, user]);
    const promptMfa = useCallback((props) => __awaiter(void 0, void 0, void 0, function* () {
        var _f;
        const methods = yield getUserMfaMethods();
        const allMethods = [
            ...methods.passkeys.map((passkey) => (Object.assign(Object.assign({}, passkey), { type: MFADeviceType.Passkey }))),
            ...methods.devices.map((device) => (Object.assign(Object.assign({}, device), { type: MFADeviceType.Totp }))),
        ];
        if (!methods.userHasVerifiedMfaMethods || allMethods.length === 0) {
            setShowAuthFlow(true);
            pushView('mfa-choose-device', {
                createMfaToken: props === null || props === void 0 ? void 0 : props.createMfaToken,
                requestedScopes: props === null || props === void 0 ? void 0 : props.requestedScopes,
            });
            return;
        }
        const lastCreatedMethod = allMethods.reduce((a, b) => {
            if (!a)
                return b;
            if (!b)
                return a;
            if (a.createdAt && b.createdAt) {
                return a.createdAt >= b.createdAt ? a : b;
            }
            if (b.createdAt) {
                return b;
            }
            return a;
        }, allMethods[0]);
        switch (lastCreatedMethod.type) {
            case MFADeviceType.Passkey: {
                const createMfaTokenOptions = (props === null || props === void 0 ? void 0 : props.createMfaToken)
                    ? { singleUse: true }
                    : undefined;
                const authenticatePasskeyPromise = authenticatePasskeyMFA$1({
                    createMfaToken: createMfaTokenOptions,
                    requestedScopes: props === null || props === void 0 ? void 0 : props.requestedScopes,
                });
                setShowAuthFlow(true);
                pushView('passkey-confirm', {
                    authenticatePasskeyPromise,
                    createMfaToken: props === null || props === void 0 ? void 0 : props.createMfaToken,
                    requestedScopes: props === null || props === void 0 ? void 0 : props.requestedScopes,
                });
                return authenticatePasskeyPromise;
            }
            case MFADeviceType.Totp: {
                if ((dynamicClient === null || dynamicClient === void 0 ? void 0 : dynamicClient.mfaToken) && !((_f = props === null || props === void 0 ? void 0 : props.requestedScopes) === null || _f === void 0 ? void 0 : _f.length)) {
                    return dynamicClient === null || dynamicClient === void 0 ? void 0 : dynamicClient.mfaToken;
                }
                const deferred = new DeferredPromise();
                setShowAuthFlow(true);
                pushView('mfa-verification', {
                    createMfaToken: props === null || props === void 0 ? void 0 : props.createMfaToken,
                    requestedScopes: props === null || props === void 0 ? void 0 : props.requestedScopes,
                });
                const mfaCompletionSuccessOff = onEvent({
                    event: 'mfaCompletionSuccess',
                    listener: ({ mfaToken }) => deferred.resolve(mfaToken),
                }, dynamicClient);
                const onAuthFlowCloseHandler = () => deferred.reject(new Error('MFA view closed'));
                dynamicEvents.on('authFlowClose', onAuthFlowCloseHandler);
                deferred.promise.finally(() => {
                    mfaCompletionSuccessOff();
                    dynamicEvents.off('authFlowClose', onAuthFlowCloseHandler);
                });
                return deferred.promise;
            }
            default:
                throw new Error('Error determining MFA method to use');
        }
    }), [
        authenticatePasskeyMFA$1,
        dynamicClient,
        getUserMfaMethods,
        pushView,
        setShowAuthFlow,
    ]);
    return {
        isStepUpRequired,
        promptMfa,
        resetState,
        sendOtp,
        state,
        verifyOtp,
        verifyPasskeyMfa,
        verifyRecoveryCode,
        verifyTotpMfa,
        verifyWallet,
    };
};

export { useStepUpAuthentication };
