// @dynamic-labs-sdk/client@0.14.0/dist/getSignedSessionId-BKKO2mqj.esm.js
import { O as getCore, g as getSessionKeys, l as createApiClient, y as assertDefined } from "./InvalidParamError-DQn49mNB.esm.js";
import { i as bufferToHex, r as consumeMfaToken, t as isMfaRequiredForAction } from "./isMfaRequiredForAction-MvVGc7k3.esm.js";

//#region src/modules/mfa/consumeMfaTokenIfRequiredForAction/consumeMfaTokenIfRequiredForAction.ts
const consumeMfaTokenIfRequiredForAction = async ({ mfaAction }, client) => {
	if (!await isMfaRequiredForAction({ mfaAction }, client)) return;
	return consumeMfaToken(client);
};

//#endregion
//#region src/utils/p256Sign/p256Sign.ts
const p256Sign = async ({ privateKeyJwk, message }) => {
	const privateKey = await crypto.subtle.importKey("jwk", privateKeyJwk, {
		name: "ECDSA",
		namedCurve: "P-256"
	}, false, ["sign"]);
	return bufferToHex(await crypto.subtle.sign({
		hash: "SHA-256",
		name: "ECDSA"
	}, privateKey, new TextEncoder().encode(message)));
};

//#endregion
//#region src/modules/sessionKeys/generateNonceSignature/generateNonceSignature.ts
const generateNonceSignature = async ({ nonce }, client) => {
	const sessionKeys = getSessionKeys(client);
	assertDefined(sessionKeys, "Session keys not found");
	const { privateKeyJwk } = sessionKeys;
	return { signature: await p256Sign({
		message: nonce,
		privateKeyJwk
	}) };
};

//#endregion
//#region src/modules/sessionKeys/generateSessionSignature/generateSessionSignature.ts
const generateSessionSignature = async ({ sessionId }, client) => {
	const sessionKeys = getSessionKeys(client);
	assertDefined(sessionKeys, "Session keys not found");
	const { privateKeyJwk } = sessionKeys;
	return { signature: await p256Sign({
		message: sessionId,
		privateKeyJwk
	}) };
};

//#endregion
//#region src/modules/sessionKeys/getSignedSessionId/getSignedSessionId.ts
const getSignedSessionId = async (client) => {
	const core = getCore(client);
	const sessionId = core.state.get().user?.sessionId;
	assertDefined(sessionId, "Session ID is required");
	const { signature: signedSessionId } = await generateSessionSignature({ sessionId }, client);
	assertDefined(signedSessionId, "Signed session ID is required");
	const { nonce } = await createApiClient({}, client).getNonce({ environmentId: core.environmentId });
	assertDefined(nonce, "Nonce is required");
	const { signature: nonceSignature } = await generateNonceSignature({ nonce }, client);
	assertDefined(nonceSignature, "Nonce signature is required");
	return { signature: `${signedSessionId}/${nonceSignature}/${nonce}` };
};

//#endregion
export { p256Sign as n, consumeMfaTokenIfRequiredForAction as r, getSignedSessionId as t };
