// @dynamic-labs-sdk/client@0.14.0/dist/NotWaasWalletAccountError-D0BXVuKy.esm.js
import { D as BaseError, O as getCore, l as createApiClient, w as getDefaultClient } from "./InvalidParamError-DQn49mNB.esm.js";
import { r as updateAuthFromVerifyResponse } from "./getVerifiedCredentialForWalletAccount-DQ7g6O7y.esm.js";

//#region src/modules/user/refreshAuth/refreshAuth.ts
/**
* Refreshes the current user's data from the server.
*
* This function fetches the latest user information and token from the backend
* updating the local user and token states with any changes.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves to the verify response.
*/
const refreshAuth = async (client = getDefaultClient()) => {
	const core = getCore(client);
	const response = await createApiClient({}, client).refreshAuth({ environmentId: core.environmentId });
	updateAuthFromVerifyResponse({ response }, client);
	return response;
};

//#endregion
//#region src/errors/NotWaasWalletAccountError.ts
var NotWaasWalletAccountError = class extends BaseError {
	constructor({ walletAddress }) {
		super({
			cause: null,
			code: "not_waas_wallet_account_error",
			docsUrl: null,
			name: "NotWaasWalletAccountError",
			shortMessage: `Wallet account ${walletAddress} is not a Dynamic WaaS wallet account`
		});
	}
};

//#endregion
export { refreshAuth as n, NotWaasWalletAccountError as t };
