// dappscout-iframe@0.4.0/dist/dappscout-iframe.es.js
var ie = Object.defineProperty;
var le = (t, r, s) => r in t ? ie(t, r, { enumerable: !0, configurable: !0, writable: !0, value: s }) : t[r] = s;
var _ = (t, r, s) => (le(t, typeof r != "symbol" ? r + "" : r, s), s);
import ue, { useState as $, useEffect as q, createContext as fe, useRef as de, useCallback as V, useContext as me } from "react";
import { createPublicClient as Ee, http as pe, getAddress as Re, isHex as _e, fromHex as ve, hexToNumber as Te } from "viem";
var z = { exports: {} }, D = {};
/**
 * @license React
 * react-jsx-runtime.production.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var te;
function he() {
  if (te)
    return D;
  te = 1;
  var t = Symbol.for("react.transitional.element"), r = Symbol.for("react.fragment");
  function s(c, i, l) {
    var p = null;
    if (l !== void 0 && (p = "" + l), i.key !== void 0 && (p = "" + i.key), "key" in i) {
      l = {};
      for (var R in i)
        R !== "key" && (l[R] = i[R]);
    } else
      l = i;
    return i = l.ref, {
      $$typeof: t,
      type: c,
      key: p,
      ref: i !== void 0 ? i : null,
      props: l
    };
  }
  return D.Fragment = r, D.jsx = s, D.jsxs = s, D;
}
var L = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ne;
function be() {
  return ne || (ne = 1, process.env.NODE_ENV !== "production" && function() {
    function t(e) {
      if (e == null)
        return null;
      if (typeof e == "function")
        return e.$$typeof === j ? null : e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case b:
          return "Fragment";
        case x:
          return "Profiler";
        case f:
          return "StrictMode";
        case o:
          return "Suspense";
        case E:
          return "SuspenseList";
        case Y:
          return "Activity";
      }
      if (typeof e == "object")
        switch (typeof e.tag == "number" && console.error(
          "Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."
        ), e.$$typeof) {
          case k:
            return "Portal";
          case S:
            return (e.displayName || "Context") + ".Provider";
          case y:
            return (e._context.displayName || "Context") + ".Consumer";
          case a:
            var n = e.render;
            return e = e.displayName, e || (e = n.displayName || n.name || "", e = e !== "" ? "ForwardRef(" + e + ")" : "ForwardRef"), e;
          case v:
            return n = e.displayName || null, n !== null ? n : t(e.type) || "Memo";
          case w:
            n = e._payload, e = e._init;
            try {
              return t(e(n));
            } catch {
            }
        }
      return null;
    }
    function r(e) {
      return "" + e;
    }
    function s(e) {
      try {
        r(e);
        var n = !1;
      } catch {
        n = !0;
      }
      if (n) {
        n = console;
        var u = n.error, d = typeof Symbol == "function" && Symbol.toStringTag && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return u.call(
          n,
          "The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",
          d
        ), r(e);
      }
    }
    function c(e) {
      if (e === b)
        return "<>";
      if (typeof e == "object" && e !== null && e.$$typeof === w)
        return "<...>";
      try {
        var n = t(e);
        return n ? "<" + n + ">" : "<...>";
      } catch {
        return "<...>";
      }
    }
    function i() {
      var e = M.A;
      return e === null ? null : e.getOwner();
    }
    function l() {
      return Error("react-stack-top-frame");
    }
    function p(e) {
      if (X.call(e, "key")) {
        var n = Object.getOwnPropertyDescriptor(e, "key").get;
        if (n && n.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function R(e, n) {
      function u() {
        Z || (Z = !0, console.error(
          "%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",
          n
        ));
      }
      u.isReactWarning = !0, Object.defineProperty(e, "key", {
        get: u,
        configurable: !0
      });
    }
    function I() {
      var e = t(this.type);
      return Q[e] || (Q[e] = !0, console.error(
        "Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release."
      )), e = this.props.ref, e !== void 0 ? e : null;
    }
    function A(e, n, u, d, C, T, G, H) {
      return u = T.ref, e = {
        $$typeof: O,
        type: e,
        key: n,
        props: T,
        _owner: C
      }, (u !== void 0 ? u : null) !== null ? Object.defineProperty(e, "ref", {
        enumerable: !1,
        get: I
      }) : Object.defineProperty(e, "ref", { enumerable: !1, value: null }), e._store = {}, Object.defineProperty(e._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: 0
      }), Object.defineProperty(e, "_debugInfo", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: null
      }), Object.defineProperty(e, "_debugStack", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: G
      }), Object.defineProperty(e, "_debugTask", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: H
      }), Object.freeze && (Object.freeze(e.props), Object.freeze(e)), e;
    }
    function N(e, n, u, d, C, T, G, H) {
      var m = n.children;
      if (m !== void 0)
        if (d)
          if (se(m)) {
            for (d = 0; d < m.length; d++)
              W(m[d]);
            Object.freeze && Object.freeze(m);
          } else
            console.error(
              "React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead."
            );
        else
          W(m);
      if (X.call(n, "key")) {
        m = t(e);
        var P = Object.keys(n).filter(function(ce) {
          return ce !== "key";
        });
        d = 0 < P.length ? "{key: someKey, " + P.join(": ..., ") + ": ...}" : "{key: someKey}", re[m + d] || (P = 0 < P.length ? "{" + P.join(": ..., ") + ": ...}" : "{}", console.error(
          `A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`,
          d,
          m,
          P,
          m
        ), re[m + d] = !0);
      }
      if (m = null, u !== void 0 && (s(u), m = "" + u), p(n) && (s(n.key), m = "" + n.key), "key" in n) {
        u = {};
        for (var U in n)
          U !== "key" && (u[U] = n[U]);
      } else
        u = n;
      return m && R(
        u,
        typeof e == "function" ? e.displayName || e.name || "Unknown" : e
      ), A(
        e,
        m,
        T,
        C,
        i(),
        u,
        G,
        H
      );
    }
    function W(e) {
      typeof e == "object" && e !== null && e.$$typeof === O && e._store && (e._store.validated = 1);
    }
    var g = ue, O = Symbol.for("react.transitional.element"), k = Symbol.for("react.portal"), b = Symbol.for("react.fragment"), f = Symbol.for("react.strict_mode"), x = Symbol.for("react.profiler"), y = Symbol.for("react.consumer"), S = Symbol.for("react.context"), a = Symbol.for("react.forward_ref"), o = Symbol.for("react.suspense"), E = Symbol.for("react.suspense_list"), v = Symbol.for("react.memo"), w = Symbol.for("react.lazy"), Y = Symbol.for("react.activity"), j = Symbol.for("react.client.reference"), M = g.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE, X = Object.prototype.hasOwnProperty, se = Array.isArray, B = console.createTask ? console.createTask : function() {
      return null;
    };
    g = {
      react_stack_bottom_frame: function(e) {
        return e();
      }
    };
    var Z, Q = {}, K = g.react_stack_bottom_frame.bind(
      g,
      l
    )(), ee = B(c(l)), re = {};
    L.Fragment = b, L.jsx = function(e, n, u, d, C) {
      var T = 1e4 > M.recentlyCreatedOwnerStacks++;
      return N(
        e,
        n,
        u,
        !1,
        d,
        C,
        T ? Error("react-stack-top-frame") : K,
        T ? B(c(e)) : ee
      );
    }, L.jsxs = function(e, n, u, d, C) {
      var T = 1e4 > M.recentlyCreatedOwnerStacks++;
      return N(
        e,
        n,
        u,
        !0,
        d,
        C,
        T ? Error("react-stack-top-frame") : K,
        T ? B(c(e)) : ee
      );
    };
  }()), L;
}
process.env.NODE_ENV === "production" ? z.exports = he() : z.exports = be();
var ge = z.exports;
const ae = () => "7.6.0", we = (t) => t.toString(16).padStart(2, "0"), Ce = (t) => {
  const r = new Uint8Array((t || 40) / 2);
  return window.crypto.getRandomValues(r), Array.from(r, we).join("");
}, Ae = () => typeof window < "u" ? Ce(10) : (/* @__PURE__ */ new Date()).getTime().toString(36);
class F {
}
_(F, "makeRequest", (r, s) => ({
  id: Ae(),
  method: r,
  params: s,
  env: {
    sdkVersion: ae()
  }
})), _(F, "makeResponse", (r, s, c) => ({
  id: r,
  success: !0,
  version: c,
  data: s
})), _(F, "makeErrorResponse", (r, s, c) => ({
  id: r,
  success: !1,
  error: s,
  version: c
}));
var h = /* @__PURE__ */ ((t) => (t.sendTransactions = "sendTransactions", t.rpcCall = "rpcCall", t.getChainInfo = "getChainInfo", t.getSafeInfo = "getSafeInfo", t.getTxBySafeTxHash = "getTxBySafeTxHash", t.getSafeBalances = "getSafeBalances", t.signMessage = "signMessage", t.signTypedMessage = "signTypedMessage", t.getEnvironmentInfo = "getEnvironmentInfo", t.requestAddressBook = "requestAddressBook", t.wallet_getPermissions = "wallet_getPermissions", t.wallet_requestPermissions = "wallet_requestPermissions", t.wallet_switchEthereumChain = "wallet_switchEthereumChain", t))(h || {}), J = /* @__PURE__ */ ((t) => (t.AWAITING_CONFIRMATIONS = "AWAITING_CONFIRMATIONS", t.AWAITING_EXECUTION = "AWAITING_EXECUTION", t.CANCELLED = "CANCELLED", t.FAILED = "FAILED", t.SUCCESS = "SUCCESS", t.PENDING = "PENDING", t.WILL_BE_REPLACED = "WILL_BE_REPLACED", t))(J || {});
class ye {
  constructor(r) {
    _(this, "iframeRef");
    _(this, "handlers", /* @__PURE__ */ new Map());
    _(this, "on", (r, s) => {
      this.handlers.set(r, s);
    });
    _(this, "isValidMessage", (r) => {
      var i;
      if (Object.prototype.hasOwnProperty.call(r.data, "isCookieEnabled"))
        return !0;
      const s = ((i = this.iframeRef.current) == null ? void 0 : i.contentWindow) === r.source, c = Object.values(h).includes(r.data.method);
      return s && c;
    });
    _(this, "canHandleMessage", (r) => !!this.handlers.get(r.data.method));
    _(this, "send", (r, s, c = !1) => {
      var p, R;
      const i = ae(), l = c ? F.makeErrorResponse(
        s,
        r,
        i
      ) : F.makeResponse(s, r, i);
      (R = (p = this.iframeRef.current) == null ? void 0 : p.contentWindow) == null || R.postMessage(l, "*");
    });
    _(this, "handleIncomingMessage", async (r) => {
      const s = this.isValidMessage(r), c = this.canHandleMessage(r);
      if (s && c) {
        r.data.method === "rpcCall" && // eslint-disable-next-line @typescript-eslint/no-explicit-any
        r.data.params.call === "eth_getBlockByNumber" || console.log("*", r.data);
        const i = this.handlers.get(r.data.method);
        try {
          const l = await i(r);
          typeof l < "u" && this.send(l, r.data.id);
        } catch (l) {
          this.send(l.message, r.data.id, !0);
        }
      }
    });
    _(this, "clear", () => {
      window.removeEventListener("message", this.handleIncomingMessage);
    });
    this.iframeRef = r, window.addEventListener("message", this.handleIncomingMessage);
  }
}
const Se = (t) => {
  const [r, s] = $(
    void 0
  );
  return q(() => {
    let c;
    return ((l) => {
      c = new ye(l), s(c);
    })(t), () => {
      c == null || c.clear();
    };
  }, [t]), r;
}, oe = fe({
  iframeRef: null,
  isReady: !1
}), Pe = ({
  children: t,
  address: r,
  appUrl: s,
  chainId: c,
  rpcUrl: i,
  sendTransaction: l,
  signMessage: p,
  signTypedData: R,
  switchChain: I
}) => {
  const [A, N] = $(), [W, g] = $(!1), [O, k] = $(!0), b = de(null), f = Se(b), x = V(
    function(a, o) {
      var v, w;
      const E = {
        ...a,
        requestId: o || Math.trunc(window.performance.now()),
        version: "0.4.2"
      };
      b && ((w = (v = b.current) == null ? void 0 : v.contentWindow) == null || w.postMessage(
        E,
        s
      ));
    },
    [b, s]
  ), y = V((a, o) => {
    x({ messageId: "TRANSACTION_CONFIRMED", data: a }, o), f == null || f.send(a, o);
  }, [f, x]), S = V((a) => {
    x({ messageId: "TRANSACTION_REJECTED", data: {} }, a), f == null || f.send("Transaction was rejected", a, !0);
  }, [f, x]);
  return q(() => {
    if (!i || !c) {
      N(void 0), k(!0);
      return;
    }
    const a = Ee({ transport: pe(i) });
    N(a), k(!1), (async () => {
      try {
        const o = await a.getChainId();
        o !== c && (console.error(
          `[DappscoutIframeProvider] Provided chainId ${c} differs from RPC chainId ${o}.`
        ), k(!0));
      } catch (o) {
        console.error("[DappscoutIframeProvider] Failed to validate chainId", o), k(!0);
      }
    })();
  }, [i, c]), q(() => {
    if (!A || !f || !c || O) {
      g(!1);
      return;
    }
    return f.on(h.getSafeInfo, async () => ({
      safeAddress: r,
      chainId: c,
      owners: [],
      threshold: 1,
      isReadOnly: !1
    })), f.on(h.getEnvironmentInfo, async () => ({
      origin: document.location.origin
    })), f.on(h.rpcCall, async (a) => {
      console.log("communicator.rpcCall", a);
      const o = a.data.params;
      try {
        return await A.request({
          method: o.call,
          params: o.params
        });
      } catch (E) {
        return E;
      }
    }), f.on(h.getTxBySafeTxHash, async (a) => {
      console.log("communicator.getTxBySafeTxHash", a);
      const { safeTxHash: o } = a.data.params, { blockNumber: E, to: v, input: w, value: Y } = await A.getTransaction({ hash: o }), { timestamp: j } = await A.getBlock({ blockNumber: E });
      return {
        txId: o,
        txStatus: E ? J.SUCCESS : J.PENDING,
        executedAt: Number(j),
        txInfo: {
          type: "Custom",
          to: {
            value: v ?? ""
          },
          dataSize: w,
          value: Y.toString(),
          isCancellation: !1
        },
        txHash: o
      };
    }), f.on(h.sendTransactions, async (a) => {
      console.log("communicator.sendTransactions", a);
      try {
        const o = a.data.params.txs.map(
          // remove maxFeePerGas and maxPriorityFeePerGas from txs because they cause an error sometimes
          ({ to: v, maxFeePerGas: w, maxPriorityFeePerGas: Y, ...j }) => ({
            // eslint-disable-line @typescript-eslint/no-unused-vars
            to: v ? Re(v) : null,
            // checksummed
            ...j
          })
        ), E = await (l == null ? void 0 : l(o[0]));
        y({ safeTxHash: E }, a.data.id);
      } catch {
        S(a.data.id);
      }
    }), f.on(h.signMessage, async (a) => {
      console.log("communicator.signMessage", a);
      try {
        const { message: o } = a.data.params, E = _e(o) ? ve(o, "string") : o, v = await (p == null ? void 0 : p(E));
        y({ signature: v }, a.data.id);
      } catch {
        S(a.data.id);
      }
    }), f.on(h.signTypedMessage, async (a) => {
      console.log("communicator.signTypedMessage", a);
      try {
        const { typedData: o } = a.data.params, E = await (R == null ? void 0 : R(o));
        y({ signature: E }, a.data.id);
      } catch {
        S(a.data.id);
      }
    }), f.on(h.wallet_switchEthereumChain, async (a) => {
      try {
        const { chainId: o } = a.data.params;
        await (I == null ? void 0 : I(Te(o))), y({ chainId: o }, a.data.id);
      } catch {
        S(a.data.id);
      }
    }), g(!0), () => {
      g(!1);
    };
  }, [
    f,
    r,
    A,
    y,
    S,
    l,
    p,
    R,
    I,
    c,
    O
  ]), /* @__PURE__ */ ge.jsx(oe.Provider, { value: { iframeRef: b, isReady: W }, children: t });
}, Ne = () => me(oe);
export {
  Pe as DappscoutIframeProvider,
  Ne as useDappscoutIframe
};
