// @dynamic-labs/sdk-react-core@4.67.2/src/lib/utils/hooks/useClientSessionKeys/useClientSessionKeys.js
'use client'
import { __awaiter } from '../../../../../_virtual/_tslib.js';
import { useRef, useCallback } from 'react';
import { StorageService } from '@dynamic-labs/utils';
import '@dynamic-labs-sdk/client/core';
import '@dynamic-labs/sdk-api-core';
import '../../../client/client.js';
import '@dynamic-labs-sdk/client';
import '../../../config/ApiEndpoint.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import 'react/jsx-runtime';
import '../../../context/ViewContext/ViewContext.js';
import { logger } from '../../../shared/logger.js';
import '@dynamic-labs/wallet-book';
import '../../constants/colors.js';
import '../../constants/values.js';
import '../../../shared/consts/index.js';
import '@dynamic-labs/multi-wallet';
import 'react-international-phone';
import '../../../store/state/nonce/nonce.js';
import '@dynamic-labs/locale';
import '../../../store/state/dynamicContextProps/dynamicContextProps.js';
import '../../../store/state/primaryWalletId/primaryWalletId.js';
import '../../../store/state/connectedWalletsInfo/connectedWalletsInfo.js';
import { getClientSessionPublicKey, getClientSessionSignature, getClientSessionNonceSignature } from '../../functions/clientSessionKeys/getClientSessionKeys.js';
import '../../functions/getWaasAddressTypeLabel/getWaasAddressTypeLabel.js';
import '../../../events/dynamicEvents.js';
import { getUserProfile } from '../../../client/extension/user/getUserProfile/getUserProfile.js';
import { useOnboardingCompleteUser } from '../../../client/extension/user/useOnboardingCompleteUser/useOnboardingCompleteUser.js';
import { getMinAuthToken } from '../../../client/extension/functions/getMinAuthToken/getMinAuthToken.js';
import { CLIENT_SESSION_KEYS, CLIENT_SESSION_KEYS_STORAGE_OPTIONS } from '../../constants/sessionStorage.js';

const useClientSessionKeys = () => {
    const user = useOnboardingCompleteUser();
    const sessionPublicKey = getClientSessionPublicKey();
    // Use refs to capture latest values without causing function reference changes
    const userRef = useRef(user);
    const sessionPublicKeyRef = useRef(sessionPublicKey);
    // Update refs on every render
    userRef.current = user;
    sessionPublicKeyRef.current = sessionPublicKey;
    /**
     * Get a signed session ID with nonce for replay protection.
     * Returns signature in format: `${signedSessionId}/${signedNonce}/${nonce}`
     *
     * This is a lazy function - nonce API is only called when this function is invoked.
     */
    const getSignedSessionId = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        const currentUser = userRef.current;
        const currentPublicKey = sessionPublicKeyRef.current;
        if (!(currentUser === null || currentUser === void 0 ? void 0 : currentUser.sessionId) || !currentPublicKey) {
            logger.instrument('[SessionChaining] Session ID and public key are required', {
                hasClientSessionKeysInSecureStorage: Boolean(StorageService.getItem(CLIENT_SESSION_KEYS, CLIENT_SESSION_KEYS_STORAGE_OPTIONS)),
                hasLocalStorage: Boolean(StorageService.getSourceStorage('localStorage')),
                hasSecureStorage: Boolean(StorageService.getSourceStorage('secureStorage')),
                hasSessionId: currentUser ? Boolean(currentUser.sessionId) : false,
                hasSessionPublicKey: Boolean(currentPublicKey),
                hasSessionStorage: Boolean(StorageService.getSourceStorage('sessionStorage')),
                hasUser: Boolean(currentUser),
                key: 'session_chaining_missing_requirements',
                sessionId: currentUser === null || currentUser === void 0 ? void 0 : currentUser.sessionId,
                sessionPublicKey: currentPublicKey,
                stateHasMinAuthToken: Boolean(getMinAuthToken()),
                stateHasUser: Boolean(getUserProfile()),
                time: 0,
                userId: currentUser === null || currentUser === void 0 ? void 0 : currentUser.id,
            });
            throw new Error('Session ID and public key are required');
        }
        const { sessionSignature: signedSessionId } = yield getClientSessionSignature({
            sessionId: currentUser.sessionId,
        });
        const { nonce, nonceSignature } = yield getClientSessionNonceSignature({
            environmentId: currentUser.projectEnvironmentId,
        });
        const sessionSignature = `${signedSessionId}/${nonceSignature}/${nonce}`;
        logger.instrument('[SessionChaining] Generated session signature', {
            environmentId: currentUser.projectEnvironmentId,
            key: 'generated_session_signature',
            nonce,
            nonceSignature,
            sessionId: currentUser.sessionId,
            sessionPublicKey: currentPublicKey,
            sessionSignature,
            signedSessionId,
            time: 0,
        });
        return sessionSignature;
    }), []);
    return {
        getSignedSessionId,
        sessionPublicKey,
    };
};

export { useClientSessionKeys };
