// @reown/appkit-scaffold-ui@1.7.20/dist/esm/src/views/w3m-deposit-from-exchange-view/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, html } from 'lit';
import { state } from 'lit/decorators.js';
import { AssetUtil, ChainController, ExchangeController, RouterController, SnackController } from '@reown/appkit-controllers';
import { customElement } from '@reown/appkit-ui';
import '@reown/appkit-ui/wui-button';
import '@reown/appkit-ui/wui-flex';
import '@reown/appkit-ui/wui-icon-link';
import '@reown/appkit-ui/wui-image';
import '@reown/appkit-ui/wui-list-item';
import '@reown/appkit-ui/wui-shimmer';
import '@reown/appkit-ui/wui-text';
import styles from './styles.js';
const PRESET_AMOUNTS = [10, 50, 100];
let W3mDepositFromExchangeView = class W3mDepositFromExchangeView extends LitElement {
    constructor() {
        super();
        this.unsubscribe = [];
        this.network = ChainController.state.activeCaipNetwork;
        this.exchanges = ExchangeController.state.exchanges;
        this.isLoading = ExchangeController.state.isLoading;
        this.amount = ExchangeController.state.amount;
        this.tokenAmount = ExchangeController.state.tokenAmount;
        this.priceLoading = ExchangeController.state.priceLoading;
        this.isPaymentInProgress = ExchangeController.state.isPaymentInProgress;
        this.currentPayment = ExchangeController.state.currentPayment;
        this.paymentId = ExchangeController.state.paymentId;
        this.unsubscribe.push(ExchangeController.subscribe(exchangeState => {
            this.exchanges = exchangeState.exchanges;
            this.isLoading = exchangeState.isLoading;
            this.amount = exchangeState.amount;
            this.tokenAmount = exchangeState.tokenAmount;
            this.priceLoading = exchangeState.priceLoading;
            this.paymentId = exchangeState.paymentId;
            this.isPaymentInProgress = exchangeState.isPaymentInProgress;
            this.currentPayment = exchangeState.currentPayment;
            const shouldHandlePaymentInProgress = exchangeState.isPaymentInProgress &&
                exchangeState.currentPayment?.exchangeId &&
                exchangeState.currentPayment?.sessionId &&
                exchangeState.paymentId;
            if (shouldHandlePaymentInProgress) {
                this.handlePaymentInProgress();
            }
        }));
    }
    disconnectedCallback() {
        this.unsubscribe.forEach(unsubscribe => unsubscribe());
        ExchangeController.reset();
    }
    firstUpdated() {
        ExchangeController.fetchExchanges();
        ExchangeController.fetchTokenPrice();
    }
    render() {
        return html `
      <wui-flex flexDirection="column" gap="xs" class="container">
        ${this.amountInputTemplate()} ${this.exchangesTemplate()}
      </wui-flex>
    `;
    }
    exchangesTemplate() {
        return html `
      <wui-flex
        flexDirection="column"
        gap="xs"
        .padding=${['xs', 's', 's', 's']}
        class="exchanges-container"
      >
        ${this.exchanges.map(exchange => html `<wui-list-item
              @click=${() => this.onExchangeClick(exchange)}
              chevron
              variant="image"
              imageSrc=${exchange.imageUrl}
              ?loading=${this.isLoading}
              ?disabled=${!this.amount}
            >
              <wui-text variant="paragraph-500" color="fg-200">
                Deposit from ${exchange.name}
              </wui-text>
            </wui-list-item>`)}
      </wui-flex>
    `;
    }
    amountInputTemplate() {
        return html `
      <wui-flex flexDirection="column" gap="s" .padding=${['0', 's', 's', 's']} class="amount-input-container">
        <wui-flex justifyContent="space-between">
          <wui-text variant="paragraph-500" color="fg-200">Asset</wui-text>
          <wui-chip-button
            data-testid="deposit-from-exchange-asset-button"
            text=${this.network?.nativeCurrency.symbol || ''}
            imageSrc=${AssetUtil.getNetworkImage(this.network)}
            size="sm"
            variant="gray"
            icon=${null}
          ></wui-chip-button>
        </wui-flex>
        <wui-flex flexDirection="column" alignItems="center" justifyContent="center">
          <wui-flex alignItems="center" gap="4xs">
            <wui-text variant="2xl-500" color="fg-200">${this.amount}</wui-text>
            <wui-text variant="paragraph-500" color="fg-200">USD</wui-text>
          </wui-flex>
          ${this.tokenAmountTemplate()}
          </wui-flex>
          <wui-flex justifyContent="space-between" gap="xs">
            ${PRESET_AMOUNTS.map(amount => html `<wui-button @click=${() => this.onPresetAmountClick(amount)} variant=${this.amount === amount ? 'accent' : 'shade'} size="sm" fullWidth>$${amount}</wui-button>`)}
          </wui-flex>
        </wui-flex>
      </wui-flex>
    `;
    }
    tokenAmountTemplate() {
        if (this.priceLoading) {
            return html `<wui-shimmer
        width="65px"
        height="20px"
        borderRadius="xxs"
        variant="light"
      ></wui-shimmer>`;
        }
        return html `
      <wui-text variant="paragraph-500" color="fg-200">
        ${this.tokenAmount} ${this.network?.nativeCurrency.symbol}
      </wui-text>
    `;
    }
    async onExchangeClick(exchange) {
        if (this.amount) {
            await ExchangeController.handlePayWithExchange(exchange.id);
        }
    }
    handlePaymentInProgress() {
        if (this.isPaymentInProgress &&
            this.currentPayment?.exchangeId &&
            this.currentPayment?.sessionId &&
            this.paymentId) {
            SnackController.showLoading('Deposit in progress...');
            RouterController.replace('Account');
            ExchangeController.waitUntilComplete({
                exchangeId: this.currentPayment.exchangeId,
                sessionId: this.currentPayment.sessionId,
                paymentId: this.paymentId
            }).then(status => {
                if (status.status === 'SUCCESS') {
                    SnackController.showSuccess('Deposit completed');
                }
                else if (status.status === 'FAILED') {
                    SnackController.showError('Deposit failed');
                }
            });
        }
    }
    onPresetAmountClick(amount) {
        ExchangeController.setAmount(amount);
    }
};
W3mDepositFromExchangeView.styles = styles;
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "network", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "exchanges", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "isLoading", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "amount", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "tokenAmount", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "priceLoading", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "isPaymentInProgress", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "currentPayment", void 0);
__decorate([
    state()
], W3mDepositFromExchangeView.prototype, "paymentId", void 0);
W3mDepositFromExchangeView = __decorate([
    customElement('w3m-deposit-from-exchange-view')
], W3mDepositFromExchangeView);
export { W3mDepositFromExchangeView };
