// @dynamic-labs/sdk-react-core@4.67.2/src/lib/utils/constants/values.js
'use client'
import { ChainEnum } from '@dynamic-labs/sdk-api-core';

const DEFAULT_NUMBER_OF_WALLETS_TO_SHOW = 10;
// references:
// https://www.iana.org/assignments/jwt/jwt.xhtml#claims
// https://github.com/jaredhanson/id-blockchain-claims-in-jwt/blob/main/draft.md
const ChainEnumToVerifiedCredentialName = {
    ALEO: 'aleo',
    ALGO: 'algorand',
    APTOS: 'aptos',
    BTC: 'bip122', // see: https://namespaces.chainagnostic.org/bip122/caip2
    COSMOS: 'cosmos',
    ETH: 'eip155',
    EVM: 'eip155',
    FLOW: 'flow',
    SOL: 'solana',
    STARK: 'starknet',
    STELLAR: 'stellar',
    SUI: 'sui',
    TON: 'ton',
    TRON: 'tron',
};
const VerifiedCredentialNameToChainEnum = {
    aleo: ChainEnum.Aleo,
    algorand: ChainEnum.Algo,
    aptos: ChainEnum.Aptos,
    bip122: ChainEnum.Btc,
    cosmos: ChainEnum.Cosmos,
    eip155: ChainEnum.Evm,
    flow: ChainEnum.Flow,
    solana: ChainEnum.Sol,
    starknet: ChainEnum.Stark,
    stellar: ChainEnum.Stellar,
    sui: ChainEnum.Sui,
    ton: ChainEnum.Ton,
    tron: ChainEnum.Tron,
};

export { ChainEnumToVerifiedCredentialName, DEFAULT_NUMBER_OF_WALLETS_TO_SHOW, VerifiedCredentialNameToChainEnum };
