// @dynamic-labs/sdk-react-core@4.67.2/src/lib/shared/assets/setup-password-terms-illustration.js
'use client'
import * as React from 'react';

var _g, _defs;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgSetupPasswordTermsIllustration = function SvgSetupPasswordTermsIllustration(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 317,
    height: 85,
    viewBox: "0 0 317 85",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React.createElement("title", {
    id: titleId
  }, title) : null, _g || (_g = /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#setup-password-terms-illustration_svg__a)"
  }, /*#__PURE__*/React.createElement("circle", {
    cx: 158.41,
    cy: 110.114,
    r: 110.114,
    fill: "url(#setup-password-terms-illustration_svg__b)",
    fillOpacity: 0.12
  }), /*#__PURE__*/React.createElement("path", {
    d: "M230.37 18.352a3.864 3.864 0 0 1 3.864 3.863v7.728a3.863 3.863 0 0 1-3.864 3.863h-38.636a3.864 3.864 0 0 1-3.864-3.863v-7.728a3.864 3.864 0 0 1 3.864-3.863h38.636Zm-32.002 5.152c-1.449 0-2.625 1.153-2.625 2.576 0 1.422 1.176 2.575 2.625 2.575s2.624-1.153 2.624-2.575c0-1.423-1.175-2.576-2.624-2.576Zm8.747 0c-1.449 0-2.624 1.153-2.624 2.576 0 1.422 1.175 2.575 2.624 2.575 1.45 0 2.625-1.153 2.625-2.575 0-1.423-1.175-2.576-2.625-2.576Zm8.749 0c-1.449 0-2.624 1.153-2.624 2.576 0 1.422 1.175 2.575 2.624 2.575 1.45 0 2.625-1.153 2.625-2.575 0-1.423-1.175-2.576-2.625-2.576Zm8.746 0c-1.449 0-2.624 1.154-2.624 2.576l.014.263c.134 1.298 1.252 2.312 2.61 2.312 1.359 0 2.477-1.014 2.612-2.313l.013-.262c0-1.423-1.175-2.576-2.625-2.576Zm0 .965c.933 0 1.659.739 1.66 1.61-.001.873-.727 1.61-1.66 1.61-.932 0-1.658-.738-1.658-1.61 0-.871.726-1.61 1.658-1.61ZM97.995 25.218c8.244-2.209 16.719 2.683 18.928 10.927 2.209 8.245-2.684 16.72-10.928 18.929-8.245 2.209-16.719-2.684-18.928-10.929-2.209-8.244 2.684-16.718 10.928-18.927Zm1.839 6.007a3.9 3.9 0 0 0-2.757 4.776l.504 1.882a6.498 6.498 0 1 0 7.532-2.018l-.504-1.882a3.9 3.9 0 0 0-4.58-2.805l-.195.047Zm2.145 7.925a1.69 1.69 0 0 1 1.532 2.919l.355 1.325a.725.725 0 1 1-1.4.375l-.355-1.325a1.69 1.69 0 0 1-.132-3.294Zm-1.395-5.126a1.002 1.002 0 0 1 1.226.709l.188.701a6.583 6.583 0 0 0-1.935.519l-.188-.702a1.003 1.003 0 0 1 .709-1.227ZM165.171 39.28a4.508 4.508 0 0 1 4.508 4.507h-27.045a8.992 8.992 0 0 0-6.618 2.893c.762-4.209 4.444-7.4 8.871-7.4h20.284ZM180.947 57.31a9.015 9.015 0 0 0-9.015-9.015h-27.046a9.015 9.015 0 0 0-9.015 9.015v18.03a9.015 9.015 0 0 0 9.015 9.015h27.046a9.015 9.015 0 0 0 9.015-9.015V57.31Z",
    fill: "#4779FF"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M155.879 66.785a7.284 7.284 0 0 0 1.936 2.16l-2.734 2.734 2.031 2.031-2.049 2.05-2.031-2.032-1.991 1.992-2.049-2.049 6.887-6.886Zm6.153-9.555a5.796 5.796 0 1 1-.001 11.591 5.796 5.796 0 0 1 .001-11.591Zm0 2.898a2.899 2.899 0 0 0 0 5.795 2.899 2.899 0 0 0 0-5.796Z",
    fill: "#fff"
  }))), _defs || (_defs = /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "setup-password-terms-illustration_svg__b",
    x1: 158.41,
    y1: 0,
    x2: 158.41,
    y2: 90.313,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#4779FF"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#4779FF",
    stopOpacity: 0
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "setup-password-terms-illustration_svg__a"
  }, /*#__PURE__*/React.createElement("rect", {
    width: 316.818,
    height: 85,
    rx: 11.591,
    fill: "#fff"
  })))));
};

export { SvgSetupPasswordTermsIllustration as ReactComponent };
