// @dynamic-labs/sdk-api-core@0.0.881/src/models/CheckoutExecutionStateEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var CheckoutExecutionStateEnum;
(function (CheckoutExecutionStateEnum) {
    CheckoutExecutionStateEnum["Initiated"] = "initiated";
    CheckoutExecutionStateEnum["SourceAttached"] = "source_attached";
    CheckoutExecutionStateEnum["Quoted"] = "quoted";
    CheckoutExecutionStateEnum["Signing"] = "signing";
    CheckoutExecutionStateEnum["Broadcasted"] = "broadcasted";
    CheckoutExecutionStateEnum["SourceConfirmed"] = "source_confirmed";
    CheckoutExecutionStateEnum["Cancelled"] = "cancelled";
    CheckoutExecutionStateEnum["Expired"] = "expired";
    CheckoutExecutionStateEnum["Failed"] = "failed";
})(CheckoutExecutionStateEnum || (CheckoutExecutionStateEnum = {}));
function CheckoutExecutionStateEnumFromJSON(json) {
    return CheckoutExecutionStateEnumFromJSONTyped(json);
}
function CheckoutExecutionStateEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CheckoutExecutionStateEnumToJSON(value) {
    return value;
}

export { CheckoutExecutionStateEnum, CheckoutExecutionStateEnumFromJSON, CheckoutExecutionStateEnumFromJSONTyped, CheckoutExecutionStateEnumToJSON };
