// @dynamic-labs/sdk-api-core@0.0.881/src/models/CheckoutSettlementStateEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var CheckoutSettlementStateEnum;
(function (CheckoutSettlementStateEnum) {
    CheckoutSettlementStateEnum["None"] = "none";
    CheckoutSettlementStateEnum["Routing"] = "routing";
    CheckoutSettlementStateEnum["Bridging"] = "bridging";
    CheckoutSettlementStateEnum["Swapping"] = "swapping";
    CheckoutSettlementStateEnum["Settling"] = "settling";
    CheckoutSettlementStateEnum["Completed"] = "completed";
    CheckoutSettlementStateEnum["Failed"] = "failed";
})(CheckoutSettlementStateEnum || (CheckoutSettlementStateEnum = {}));
function CheckoutSettlementStateEnumFromJSON(json) {
    return CheckoutSettlementStateEnumFromJSONTyped(json);
}
function CheckoutSettlementStateEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CheckoutSettlementStateEnumToJSON(value) {
    return value;
}

export { CheckoutSettlementStateEnum, CheckoutSettlementStateEnumFromJSON, CheckoutSettlementStateEnumFromJSONTyped, CheckoutSettlementStateEnumToJSON };
