// @dynamic-labs/sdk-api-core@0.0.881/src/models/CheckoutSourceTypeEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var CheckoutSourceTypeEnum;
(function (CheckoutSourceTypeEnum) {
    CheckoutSourceTypeEnum["Wallet"] = "wallet";
    CheckoutSourceTypeEnum["Exchange"] = "exchange";
    CheckoutSourceTypeEnum["DepositAddress"] = "deposit_address";
})(CheckoutSourceTypeEnum || (CheckoutSourceTypeEnum = {}));
function CheckoutSourceTypeEnumFromJSON(json) {
    return CheckoutSourceTypeEnumFromJSONTyped(json);
}
function CheckoutSourceTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CheckoutSourceTypeEnumToJSON(value) {
    return value;
}

export { CheckoutSourceTypeEnum, CheckoutSourceTypeEnumFromJSON, CheckoutSourceTypeEnumFromJSONTyped, CheckoutSourceTypeEnumToJSON };
