// @dynamic-labs/sdk-api-core@0.0.881/src/models/CheckoutRiskStateEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var CheckoutRiskStateEnum;
(function (CheckoutRiskStateEnum) {
    CheckoutRiskStateEnum["Unknown"] = "unknown";
    CheckoutRiskStateEnum["Pending"] = "pending";
    CheckoutRiskStateEnum["Cleared"] = "cleared";
    CheckoutRiskStateEnum["Blocked"] = "blocked";
    CheckoutRiskStateEnum["Review"] = "review";
})(CheckoutRiskStateEnum || (CheckoutRiskStateEnum = {}));
function CheckoutRiskStateEnumFromJSON(json) {
    return CheckoutRiskStateEnumFromJSONTyped(json);
}
function CheckoutRiskStateEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CheckoutRiskStateEnumToJSON(value) {
    return value;
}

export { CheckoutRiskStateEnum, CheckoutRiskStateEnumFromJSON, CheckoutRiskStateEnumFromJSONTyped, CheckoutRiskStateEnumToJSON };
