// @dynamic-labs/sdk-api-core@0.0.881/src/models/SettlementStrategyEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var SettlementStrategyEnum;
(function (SettlementStrategyEnum) {
    SettlementStrategyEnum["Cheapest"] = "cheapest";
    SettlementStrategyEnum["Fastest"] = "fastest";
    SettlementStrategyEnum["PreferredOrder"] = "preferred_order";
})(SettlementStrategyEnum || (SettlementStrategyEnum = {}));
function SettlementStrategyEnumFromJSON(json) {
    return SettlementStrategyEnumFromJSONTyped(json);
}
function SettlementStrategyEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function SettlementStrategyEnumToJSON(value) {
    return value;
}

export { SettlementStrategyEnum, SettlementStrategyEnumFromJSON, SettlementStrategyEnumFromJSONTyped, SettlementStrategyEnumToJSON };
